### ISSUE TITLE
`jnp.searchsorted' returns 32-bit integer when `jax_enable_x64 = True`

### ISSUE THREAD

**Issue by btalamini** [google/jax#11375]
Am I correct to expect `searchsorted` to return an `int64` when `jax_enable_x64 = True`? The other search and sort functions (such as `argmax`) use `int64` in this case:

```python
import jax
jax.config.update("jax_enable_x64", True)
from jax import numpy as jnp

a = jnp.arange(6)
print(a.dtype) # int64

i = jnp.argmax(a)
print(i.dtype) # int64

j = jnp.searchsorted(a, 4)
print(j.dtype) # int32
```

**Comment by jakevdp** [comment]
The desired behavior here is ambiguous – the x64 flag will be going away soon, and many users will not want 64-bit computations silently generated for performance reasons. Currently we handle this ambiguity by only returning 64-bit integers when the size of the array requires it: https://github.com/google/jax/blob/da5385f235c92f9cc11e3993ac46d4fdd76e668e/jax/_src/numpy/lax_numpy.py#L4355

What do you think? Is it a problem to return 32-bit indices when such indices are large enough to handle all valid outputs?

**Comment by btalamini** [comment]
> What do you think? Is it a problem to return 32-bit indices when such indices are large enough to handle all valid outputs?

No, seems reasonable. My main concern is that the types returned and consumed by indexing operations be compatible so I don't need to think about them. For instance, the following simple expression doesn't work with the x64 flag enabled:
```python
jax.lax.cond(np.max(a) < 6, lambda p: p, lambda p: 0, np.searchsorted(a, 3))
``` 
The problem being that the `0` in the false branch gets converted into a weak-typed 64-bit integer. So I have to write something more verbose, like this:
```python
jax.lax.cond(np.max(a) < 6, lambda p: p, lambda p: np.int32(0), np.searchsorted(a, 3))
```
I admit the resulting error in the first case is clear enough, and it's not *that* hard to explicitly cast to get around this. On the other hand, this kind of mental focus on typing diminishes the advantage of working in Python. It gets worse if the various searching/sorting algorithms return disparate types. 

Also, If I'm understanding #8178 correctly, the conversion of integer constants and Python integers into  64-bit will continue, so this issue would still apply in part.

**Comment by jakevdp** [comment]
Yeah, as soon as you start using `lax`-level operations you have to be much more explicit about dtypes. For the most part this is working as intended – `jax.lax` is meant to be much lower level than `jax.numpy`.

**Comment by jakevdp** [comment]
Though this is a case that could probably be remedied using some sort of fixed-point procedure like we do for `scan` and others.
